### PR TITLE
Detect integration tests in `is_in_test`

### PR DESCRIPTION
**Pull request by ClementTsang**
Related: https://github.com/rust-lang/rust-clippy/issues/13981

This PR makes it so that the `is_in_test` check accounts for if a node is in an integration test file (that is, under `tests/`). This should fix the issue where some clippy settings work for `cfg(test)` but not for integration tests.

This could be extended to `benches` as well if this seems reasonable.

---

Note that this is the first time I'm contributing to this repo, so please let me know if there's anything to change. I'm not really sure how I was meant to test this for example, especially if I want to test an example of this where it should still "fail", I guess (should I split the `integration_test` with two crates?).

changelog: [`unwrap-used`]: `allow-unwrap-in-tests` works now on integration tests.

### CODE REVIEW

**Comment by ClementTsang** [comment]
Hmmm it looks like it's causing problems because now some of _clippy's_ tests are being treated as valid under `is_in_test`... let me set this to draft for now and figure this out. For whatever reason I didn't catch this when testing locally earlier.

**Comment by blyxyas** [comment]
Hello @ClementTsang, I hope that you don't mind we using this contribution to teach @medzernik some things about reviewing (they were very interested!).

cc @medzernik Check out [this comment](https://github.com/rust-lang/rust-clippy/pull/16800#issuecomment-4701708032) first.

**Comment by ClementTsang** [comment]
@blyxyas nope, I don't mind!

**Review by medzernik — APPROVED** [review]
I didn't find anything of note.. I guess approve? :D

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/16786/98098dfe6bb203425dece84610cbcff61771a9a6..8dca46a4487d40fb70ec0984c2562ce08f5975d0)*

**Comment by blyxyas** [comment]
I'm going to review this tomorrow, sorry for the radio silence ( =ノωヽ=) (I'm finally back up!)

**Comment by ClementTsang** [comment]
No worries, thanks to both of you for taking the time to review it!

File: tests/example_integration_test/Cargo.toml
Comment (blyxyas) [inline tests/example_integration_test/Cargo.toml:1]: Could you rename this folder into `example_integration_test`?

File: tests/allow-unwrap-in-integration-tests.rs
Comment (blyxyas) [inline tests/allow-unwrap-in-integration-tests.rs:1]: This harness is waaay too complex! What about adding `example_integration_tests/tests/ui/` as a UI testing directory in compile-test? That should take care of most of this automatically (and handle all the weird edge cases, calling clippy directory is not really what we want to go through) ฅ(=^Õ ‸ Õ^=)ฅ

**Review by blyxyas — CHANGES_REQUESTED** [review]
Looking good, just a couple of asks!! 

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/16786/98098dfe6bb203425dece84610cbcff61771a9a6..8dca46a4487d40fb70ec0984c2562ce08f5975d0)*

File: tests/example_integration_test/Cargo.toml
Comment (ClementTsang) [reply tests/example_integration_test/Cargo.toml:1]: Will do!

File: tests/allow-unwrap-in-integration-tests.rs
Comment (ClementTsang) [reply tests/allow-unwrap-in-integration-tests.rs:1]: That makes sense, I wasn't totally sure on the testing part either :sweat_smile:

File: tests/example_integration_test/Cargo.toml
Comment (ClementTsang) [reply tests/example_integration_test/Cargo.toml:1]: Done.

File: tests/allow-unwrap-in-integration-tests.rs
Comment (ClementTsang) [reply tests/allow-unwrap-in-integration-tests.rs:1]: Moved it over, let me know if the way I did it makes sense!

File: tests/compile-test.rs [hunk 1]
```diff
@@ -368,6 +368,51 @@ fn run_ui_cargo(cx: &TestContext) {
...
+        vec![config],
+        |path, config| {
+            path.ends_with("Cargo.toml")
+                .then(|| ui_test::default_any_file_filter(path, config))
+        },
+        |_config, _file_contents| {},
+        Box::<dyn StatusEmitter>::from(cx.args.format),
+    )
+    .unwrap();
+}
```
Comment (blyxyas) [inline tests/compile-test.rs]: Can we merge this function with `run_ui_cargo`

**Review by blyxyas — CHANGES_REQUESTED** [review]
Everything else is looking great! Please merge those two functions and squash the commits, then we can look again (and hopefully) approve! ฅ՞•ﻌ•՞ฅ

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/16786/98098dfe6bb203425dece84610cbcff61771a9a6..d8b612201089bea5e7430c5b18cf10a030db4ff8)*

**Review by blyxyas — APPROVED** [review]
LGTM, thanks for addressing this issue! :heart: ฅ•ω•ฅ And welcome to the project!

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/16786/999866a3a7680608ef5f1937374d04548b6e26c5..1ea7460127b9c7b601bbedbd001d3972b3ba4410)*
